### DISCUSSION TITLE
baseapp 一些报错

### DISCUSSION THREAD

**Discussion by MacroGu** [kbengine/kbengine#1366] [General]
` is running ----
Logging(changed) to: D:\kbengine_learn\KBE_UE4_course_pro_server/logs/logger_baseapp.*.log

[INFO]: Found all the components!
   baseapp01: [ERROR]: KCPPacketSender::ikcp_send: send error! currPacketSize=15, ikcp_waitsnd=257, snd_wndsize=128
   baseapp01: [ERROR]: 192.168.1.195:52334: UDPPacketSender::processSend(internal, 1): Exception occurred: REASON_RESOURCE_UNAVAILABLE
   baseapp01: [ERROR]: KCPPacketSender::ikcp_send: send error! currPacketSize=15, ikcp_waitsnd=257, snd_wndsize=128
   baseapp01: [ERROR]: 192.168.1.195:52334: UDPPacketSender::processSend(internal, 2): Exception occurred: REASON_RESOURCE_UNAVAILABLE
   baseapp01: [ERROR]: KCPPacketSender::ikcp_send: send error! currPacketSize=15, ikcp_waitsnd=257, snd_wndsize=128
   baseapp01: [ERROR]: 192.168.1.195:52334: UDPPacketSender::processSend(internal, 3): Exception occurred: REASON_RESOURCE_UNAVAILABLE
   baseapp01: [ERROR]: KCPPacketSender::ikcp_send: send error! currPacketSize=15, ikcp_waitsnd=257, snd_wndsize=128
   baseapp01: [ERROR]: 192.168.1.195:52334: UDPPacketSender::processSend(internal, 4): Exception occurred: REASON_RESOURCE_UNAVAILABLE
   baseapp01: [ERROR]: KCPPacketSender::ikcp_send: send error! currPacketSize=15, ikcp_waitsnd=257, snd_wndsize=128
   baseapp01: [ERROR]: 192.168.1.195:52334: UDPPacketSender::processSend(internal, 5): Exception occurred: REASON_RESOURCE_UNAVAILABLE
   baseapp01: [ERROR]: KCPPacketSender::ikcp_send: send error! currPacketSize=15, ikcp_waitsnd=257, snd_wndsize=128
   baseapp01: [ERROR]: 192.168.1.195:52334: UDPPacketSender::processSend(internal, 6): Exception occurred: REASON_RESOURCE_UNAVAILABLE
   baseapp01: [ERROR]: KCPPacketSender::ikcp_send: send error! currPacketSize=30, ikcp_waitsnd=257, snd_wndsize=128
   baseapp01: [ERROR]: 192.168.1.195:52334: UDPPacketSender::processSend(internal, 7): Exception occurred: REASON_RESOURCE_UNAVAILABLE
   baseapp01: [ERROR]: KCPPacketSender::ikcp_send: send error! currPacketSize=41, ikcp_waitsnd=257, snd_wndsize=128
   baseapp01: [ERROR]: 192.168.1.195:52334: UDPPacketSender::processSend(internal, 8): Exception occurred: REASON_RESOURCE_UNAVAILABLE
   baseapp01: [ERROR]: KCPPacketSender::ikcp_send: send error! currPacketSize=41, ikcp_waitsnd=257, snd_wndsize=128
   baseapp01: [ERROR]: 192.168.1.195:52334: UDPPacketSender::processSend(internal, 9): Exception occurred: REASON_RESOURCE_UNAVAILABLE
   baseapp01: [ERROR]: KCPPacketSender::ikcp_send: send error! currPacketSize=41, ikcp_waitsnd=257, snd_wndsize=128
   baseapp01: [ERROR]: 192.168.1.195:52334: UDPPacketSender::processSend(external, sendfailCount(10) >= 10): Exception occurred: REASON_RESOURCE_UNAVAILABLE
   baseapp01: [ERROR]: KCPPacketSender::ikcp_send: send error! currPacketSize=15, ikcp_waitsnd=257, snd_wndsize=128
   baseapp01: [ERROR]: 192.168.1.195:52335: UDPPacketSender::processSend(internal, 1): Exception occurred: REASON_RESOURCE_UNAVAILABLE
   baseapp01: [ERROR]: KCPPacketSender::ikcp_send: send error! currPacketSize=15, ikcp_waitsnd=257, snd_wndsize=128
   baseapp01: [ERROR]: 192.168.1.195:52335: UDPPacketSender::processSend(internal, 2): Exception occurred: REASON_RESOURCE_UNAVAILABLE
   baseapp01: [ERROR]: KCPPacketSender::ikcp_send: send error! currPacketSize=15, ikcp_waitsnd=257, snd_wndsize=128
   baseapp01: [ERROR]: 192.168.1.195:52335: UDPPacketSender::processSend(internal, 3): Exception occurred: REASON_RESOURCE_UNAVAILABLE
   baseapp01: [ERROR]: KCPPacketSender::ikcp_send: send error! currPacketSize=15, ikcp_waitsnd=257, snd_wndsize=128
   baseapp01: [ERROR]: 192.168.1.195:52335: UDPPacketSender::processSend(internal, 4): Exception occurred: REASON_RESOURCE_UNAVAILABLE
   baseapp01: [ERROR]: KCPPacketSender::ikcp_send: send error! currPacketSize=15, ikcp_waitsnd=257, snd_wndsize=128
   baseapp01: [ERROR]: 192.168.1.195:52335: UDPPacketSender::processSend(internal, 5): Exception occurred: REASON_RESOURCE_UNAVAILABLE
   baseapp01: [ERROR]: KCPPacketSender::ikcp_send: send error! currPacketSize=15, ikcp_waitsnd=257, snd_wndsize=128
   baseapp01: [ERROR]: 192.168.1.195:52335: UDPPacketSender::processSend(internal, 6): Exception occurred: REASON_RESOURCE_UNAVAILABLE
   baseapp01: [ERROR]: KCPPacketSender::ikcp_send: send error! currPacketSize=30, ikcp_waitsnd=257, snd_wndsize=128
   baseapp01: [ERROR]: 192.168.1.195:52335: UDPPacketSender::processSend(internal, 7): Exception occurred: REASON_RESOURCE_UNAVAILABLE
   baseapp01: [ERROR]: KCPPacketSender::ikcp_send: send error! currPacketSize=30, ikcp_waitsnd=257, snd_wndsize=128
   baseapp01: [ERROR]: 192.168.1.195:52335: UDPPacketSender::processSend(internal, 8): Exception occurred: REASON_RESOURCE_UNAVAILABLE
   baseapp01: [ERROR]: KCPPacketSender::ikcp_send: send error! currPacketSize=44, ikcp_waitsnd=257, snd_wndsize=128
   baseapp01: [ERROR]: 192.168.1.195:52335: UDPPacketSender::processSend(internal, 9): Exception occurred: REASON_RESOURCE_UNAVAILABLE
   baseapp01: [ERROR]: KCPPacketSender::ikcp_send: send error! currPacketSize=44, ikcp_waitsnd=257, snd_wndsize=128
   baseapp01: [ERROR]: 192.168.1.195:52335: UDPPacketSender::processSend(external, sendfailCount(10) >= 10): Exception occurred: REASON_RESOURCE_UNAVAILABLE
`

请问什么原因

**Comment by kebiao** [comment]
snd_wndsize=128 发送窗口满了。  发不出去。 应该是客户端没接收或者哪里网络堵了。 或者太快

**Reply by H-J-F** [reply]
那请问该怎么解决呢？怎么样让客户端发送断开连接的消息给服务器呢？

**Reply by MacroGu** [reply]
这个是KCP，应该就是没链接

**Reply by H-J-F** [reply]
我知道是 KCP，但我推测，应该是客户端主动退出后，服务器那边还没检测到客户端已经关闭，然后一直给客户端发送消息，但因为客户端关闭，一直发送不成功，就会导致最后消息积累太多，突破消息窗口的限制，所以我就是想，如何在客户端关闭的时候，主动发消息告诉 KCP 我已经断开连接了，不然 KCP 的心跳机制就会一直发消息

**Reply by MacroGu** [reply]
应该不是这个原因， 如果按照你说的，任意一个客户端退出都会有这种情况出现， 但是并没有，这个错误只会在 服务器，客户端卡顿的时候，导致的， 应该就是上面说的，发送缓存过多，导致满了，所以报错

**Reply by H-J-F** [reply]
那我应该怎么解决这个问题呢？因为客户端都断开之后，服务器还继续发

**Reply by huhubis** [reply]
客户端和baseapp都有个disconnect方法， 你看看

**Comment by datemore111** [comment]
REASON_RESOURCE_UNAVAILABLE
